File: HW1_Task5/src/com/company/AbstractFileSystemNode.java
package com.company;

public abstract class AbstractFileSystemNode implements FileSystemNode{
    FileSystemNode parent;
    String name;
    public final FileSystemNode getParentElement(){
        return this.parent;
    }
    public String getElementName(){
        return name;
    }
    public final String getPathOf(){
        FileSystemNode parent = this.getParentElement();
        String result = this.getElementName();
        while(parent != null){
            result = parent.getElementName()+"/"+result;
            parent = parent.getParentElement();
        }
        return result;
    }
}

File: HW3_Task2/src/AbstractFileSystemNode.kt
abstract class AbstractFileSystemNode : FileSystemNode {
    var parent: FileSystemNode? = null
    var name: String = ""

    //нельзя переопределить
    final override fun getParentElement(): FileSystemNode? {
        return parent
    }

    override fun getElementName(): String {
        return name
    }

    //нельзя переопределить
    final override fun getPathOf(): String {
        var parent = getParentElement()
        var result = getElementName()
        while (parent != null) {
            result = parent.getElementName() + "/" + result
            parent = parent.getParentElement()
        }
        return result
    }
}